### PR TITLE
Fix Finder Sync Extension

### LINKED ISSUE #5755: [Bug]: Nextcloud context menu broken in OSX Big Sur
### ⚠️ Before submitting, please verify the following: ⚠️

- [X] This is a **bug**, not a question or a configuration issue.
- [X] This issue is **not** already reported on Github (I've searched it).
- [X] Nextcloud Server and Desktop Client are **up to date**. See [Server Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) and [Desktop Releases](https://nextcloud.com/install/#install-clients) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/)

### Bug description

in OSX Big Sur, NC Sync Client 3.8.2 and 3.9.50 will sync but will not provide the context menu to copy file sharing links, and the green *synced* icon signaling the sync took place is missing too.
...

### Steps to reproduce

1. Install NC Sync 3.8.2 in OSX Big Sur
2. wait for the sync ok icon in menu bar
3. go to your nextcloud synced folder
4. check if you can copy any links with a context menu (right click)

### Expected behavior

When opening the sync folder, files and folders should show a green sync icon net to the file name.
A right click on the file name would allow to copy a private or a public link to that file.
...

### Which files are affected by this bug

all synced files in the NC sync folder

### Operating system

Mac OS

### Which version of the operating system you are running.

11.7.8 Big Sur

### Package

Distro package manager

### Nextcloud Server version

25.0.7

### Nextcloud Desktop Client version

3.9.50

### Is this bug present after an update or on a fresh install?

Updated from a minor version (ex. 3.4.2 to 3.4.4)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### Are you using an external user-backend?

- [ ] Default internal user-backend
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Nextcloud Server logs

```shell
Nextclod server log here
https://wolkenmeer.com/index.php/s/eXJd4Zd5EdJJdj7
```

### Additional info

Client debug archive here
https://wolkenmeer.com/index.php/s/MPEmmWbyBsoaxP5

### PR DESCRIPTION
**Pull request by claucambra**
This PR fixes the broken Finder Sync Extension in master. Specifically:

- Fixes issues in the xcodeproj that would lead to the sytem not running the appex
- Fixes broken code in the parsing for incoming bytes over the socket with the desktop client
- Fixes the optional compile if the File Provider module

Closes #5755

### CODE REVIEW

File: shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m [hunk 1]
```diff
@@ -326,24 +326,43 @@ - (void)readFromSocket
...
+                                                             range:inBufferLengthRange];
+
+        unsigned char *buffer = _inBuffer.mutableBytes;
+        NSUInteger nullTerminatorIndex = NSUIntegerMax;
 
-            [_inBuffer replaceBytesInRange:NSMakeRange(0, firstSeparatorIndex.location + 1) withBytes:NULL length:0];
-            [_lineProcessor process:newLine];
+        if (firstSeparatorIndex.location == NSNotFound) {
+            NSLog(@"No separator found. Creating new buffer qith space for null terminator.");
+            unsigned char newBuffer[inBufferLength + 1];
```
Comment (allexzander) [inline shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m:346]: @claucambra 

`unsigned char newBuffer[inBufferLength + 1];` -> `unsigned char newBuffer[inBufferLength + 1] = {0}`

File: shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m [hunk 2]
```diff
@@ -326,24 +326,43 @@ - (void)readFromSocket
...
+            memcpy(&newBuffer, buffer, inBufferLength);
+            buffer = &newBuffer[0];
```
Comment (allexzander) [inline shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m:348]: @claucambra Once you exit the if, `unsigned char newBuffer` will get deallocated (as it is on the stack), hence `buffer` will be pointing to an invalid memory location

File: shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m [hunk 3]
```diff
@@ -326,24 +326,43 @@ - (void)readFromSocket
...
+            nullTerminatorIndex = inBufferLength;
+        } else {
+            nullTerminatorIndex = firstSeparatorIndex.location;
         }
+
+        NSAssert(nullTerminatorIndex != NSUIntegerMax, @"Null terminator index should be valid.");
+
+        buffer[nullTerminatorIndex] = 0; // Add NULL terminator, so we can use C string methods
```
Comment (allexzander) [inline shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m:356]: @claucambra beware, see my previous comment about deallocation

File: shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m [hunk 4]
```diff
@@ -326,24 +326,43 @@ - (void)readFromSocket
...
-            NSLog(@"No separator found. Stopping.");
-            return; // No separator, nope out
-        } else {
-            unsigned char *buffer = [_inBuffer mutableBytes];
-            buffer[firstSeparatorIndex.location] = 0; // Add NULL terminator, so we can use C string methods
-            
-            NSString *newLine = [NSString stringWithUTF8String:[_inBuffer bytes]];
+    NSData * const separatorData = [NSData dataWithBytes:separator length:1];
+
+    while(_inBuffer.length > 0) {
```
Comment (allexzander) [inline shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m:334]: @claucambra I am not an expert in Objective-C, but, are you sure the `_inBuffer.length` will reach zero at some point? will not we get an infinite loop?

**Review by allexzander — CHANGES_REQUESTED** [review]
@claucambra kindly check my comments

File: shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m (hunk 4 above)
Comment (claucambra) [reply shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m:334]: No, as at the end of the iteration we do `[_inBuffer replaceBytesInRange:nullTerminatorRange withBytes:NULL length:0];`

File: shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m (hunk 2 above)
Comment (claucambra) [reply shell_integration/MacOSX/NextcloudIntegration/NCDesktopClientSocketKit/LocalSocketClient.m]: What a thing to miss -- thanks! Fixed now

**Review by allexzander — APPROVED** [review]

**Comment by claucambra** [comment]
Backport attempt was a mistake, no need as file provider inclusion was reverted in stable-3.9

**Pull request merged**